lib/logger.js

```diff
@@ -1,456 +1,634 @@
+        throw new Error('Invalid hostname')
+      }
+      hostname = options.hostname
+    }
 
-  if (options.mac && !configs.MAC_ADDR_CHECK.test(options.mac)) {
-    throw new Error('Invalid MAC Address format')
-  }
+    if (has(options, 'mac')) {
+      checkStringParam(options.mac, 'MAC Address')
+      if (!constants.MAC_ADDR_CHECK.test(options.mac)) {
+        throw new Error('Invalid MAC Address format')
+      }
+      mac = options.mac
+    }
 
-  if (options.ip && !configs.IP_ADDR_CHECK.test(options.ip)) {
-    throw new Error('Invalid IP Address format')
-  }
+    if (has(options, 'ip')) {
+      checkStringParam(options.ip, 'IP Address')
+      if (!isIP(options.ip)) {
+        throw new Error('Invalid IP Address format')
+      }
+      ip = options.ip
+    }
 
-  if (options.logdna_url && !validUrl.isUri(options.logdna_url)) {
-    throw new Error('Invalid URL')
-  }
+    if (has(options, 'logdna_url')) {
+      console.warn('[Deprecated] logdna_url is deprecated.  Please use url instead.')
+      options.url = options.logdna_url
+    }
+    if (has(options, 'url')) {
+      checkStringParam(options.url, 'url')
+      if (!constants.PROTOCOL_RE.test(options.url)) {
+        const err = new Error('Invalid URL protocol')
+        err.meta = {
+          expected: 'http:// or https://'
+        }
+        throw err
+      }
+      useHttps = options.url.startsWith('https://')
+      this.url = options.url
+    }
 
-  if (options.flushLimit && Number.isInteger(options.flushLimit)) {
-    this._flushLimit = options.flushLimit
-  } else {
-    this._flushLimit = configs.FLUSH_BYTE_LIMIT
-  }
+    if (has(options, 'flushLimit')) {
+      if (!Number.isInteger(options.flushLimit)) {
+        const err = new TypeError('flushLimit must be an integer')
+        err.meta = {got: options.flushLimit}
+        throw err
+      }
+      this.flushLimit = options.flushLimit
+    }
 
-  if (options.retryTimes && Number.isInteger(options.retryTimes)) {
-    this._retryTimes = options.retryTimes
-  } else {
-    this._retryTimes = configs.RETRY_TIMES
-  }
+    if (has(options, 'flushInterval')) {
+      if (!Number.isInteger(options.flushInterval)) {
+        const err = new TypeError('flushInterval must be an integer')
+        err.meta = {got: options.flushInterval}
+        throw err
+      }
+      this.flushInterval = options.flushInterval
+    }
 
-  if (options.flushInterval && Number.isInteger(options.flushInterval)) {
-    this._flushInterval = options.flushInterval
-  } else {
-    this._flushInterval = configs.FLUSH_INTERVAL
-  }
+    if (has(options, 'baseBackoffMillis')) {
+      if (!Number.isInteger(options.baseBackoffMillis)
+        || options.baseBackoffMillis <= 0) {
 
-  if (options.retryTimeout && Number.isInteger(options.retryTimeout)) {
-    this._retryTimeout = options.retryTimeout
-  } else {
-    this._retryTimeout = configs.BACKOFF_PERIOD
-  }
+        const err = new TypeError('baseBackoffMillis must be an integer > 0')
```

Comment:
technically several of these numeric params could throw `RangeError` or `TypeError`

Reply:
yeah, `RangeError` is better here, thanks.